src/UiPath.Workflow.Runtime/Statements/Flowchart.cs

```diff
@@ -179,199 +43,166 @@ protected override void CacheMetadata(NativeActivityMetadata metadata)
-    {
-        Fx.Assert(visitNodeCallback != null, "This must be supplied since it stops us from infinitely looping.");
-
-        List<FlowNode> connected = new();
-        Stack<FlowNode> stack = new();
-        if (start == null)
+        metadata.SetChildrenCollection(new Collection<Activity>(uniqueChildren.ToList()));
+        bool WasVisited(FlowNode node)
         {
-            return;
-        }
-        stack.Push(start);
-        while (stack.Count > 0)
-        {
-            FlowNode current = stack.Pop();
-
-            if (current == null)
+            if (node is null)
             {
-                continue;
+                return true;
             }
 
-            if (visitNodeCallback(current))
+            if (node.Flowchart != null)
             {
-                connected.Clear();
-                current.GetConnectedNodes(connected);
-
-                for (int i = 0; i < connected.Count; i++)
+                if (node.Flowchart != this)
                 {
-                    stack.Push(connected[i]);
+                    metadata.AddValidationError(SR.FlowNodeCannotBeShared(node.Flowchart.DisplayName, DisplayName));
                 }
+                return true;
             }
+            return false;
         }
-    }
-
 
-    protected override void Execute(NativeActivityContext context)
-    {
-        if (StartNode != null)
+        void GatherReachableNodes()
         {
-            if (TD.FlowchartStartIsEnabled())
+            // Clear out our cached list of all nodes
+            _reachableNodes.Clear();
+            _staticBranchesByNode.Clear();
+
+            if (StartNode == null && Nodes.Count > 0)
             {
-                TD.FlowchartStart(DisplayName);
+                metadata.AddValidationError(SR.FlowchartMissingStartNode(DisplayName));
             }
-            ExecuteNodeChain(context, StartNode, null);
-        }
-        else
-        {
-            if (TD.FlowchartEmptyIsEnabled())
+            else if (StartNode is not null)
             {
-                TD.FlowchartEmpty(DisplayName);
+                DepthFirstVisitNodes();
             }
         }
-    }
 
-    private void ExecuteNodeChain(NativeActivityContext context, FlowNode node, ActivityInstance completedInstance)
-    {
-        if (node == null)
+        void DepthFirstVisitNodes()
         {
-            if (context.IsCancellationRequested)
+            Stack<FlowNode> toVisit = new();
+            toVisit.Push(StartNode);
+            GetStaticSplitsStack(StartNode).PropagateStack(StaticNodeStackInfo.EmptyStack);
+            while (toVisit.TryPop(out var current))
             {
-                Fx.Assert(completedInstance != null, "cannot request cancel if we never scheduled any children");
-                // we are done but the last child didn't complete successfully
-                if (completedInstance.State != ActivityInstanceState.Closed)
+                if (WasVisited(current))
```

Comment:
WasVisited is based on node.Flowchart != null. 
I did not understand if the new nodes are created every time before CacheMetadata

Reply:
?